yamui-screensaverd.c

```diff
@@ -0,0 +1,361 @@
+			}
+
+			if (BIT(bits[EV_ABS], ABS_MT_POSITION_X) &&
+			    BIT(bits[EV_ABS], ABS_MT_POSITION_Y)) {
+				debugf("Device supports multi-touch events.");
+				return 0;
+			}
+		} else if (BIT(bits[0], EV_KEY)) {
+			if (ioctl(fd, EVIOCGBIT(EV_KEY, KEY_MAX),
+				  bits[EV_KEY]) == -1) {
+				errorf("ioctl(, EVIOCGBIT(EV_KEY, ), ) error "
+				       "on event device");
+				return -1;
+			}
+
+			if (BIT(bits[EV_KEY], KEY_POWER)	||
+			    BIT(bits[EV_KEY], KEY_VOLUMEDOWN)	||
+			    BIT(bits[EV_KEY], KEY_VOLUMEUP)	||
+			    BIT(bits[EV_KEY], KEY_OK)		||
+			    BIT(bits[EV_KEY], KEY_ENTER)) {
+				debugf("Device supports needed key events.");
+				return 0;
+			}
+		}
+
+	debugf("Skipping unsupported device.");
+	return -1;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static void
+close_fds(int fds[], int num)
+{
+	int i;
+
+	for (i = 0; i < num; i++)
+		close(fds[i]);
+}
+
+/* ------------------------------------------------------------------------ */
+
+/* Open all /dev/input/event* files. */
+static int
+open_fds(int fds[], int *num, int max_num)
+{
+	DIR *dir;
+	struct dirent *d;
+
+	if (!(dir = opendir(DEV_INPUT_DIR))) {
+		errorf("Can't open directory %s", DEV_INPUT_DIR);
+		return -1;
+	}
+
+	while ((d = readdir(dir))) {
+		char name[256];
+
+		if (strncmp(d->d_name, EVENT_PREFIX, strlen(EVENT_PREFIX)))
+			continue; /* Not /dev/input/event* file */
+
+		snprintf(name, sizeof(name), "%s/%s", DEV_INPUT_DIR,
+			 d->d_name);
+		debugf("Processing input ivents file %s", name);
+		if ((fds[*num] = open(name, O_RDONLY)) == -1) {
+			errorf("Can't open input device %s", name);
+			continue;
+		}
+
+		if (check_device_type(fds[*num]) == -1) {
+			close(fds[*num]);
+			continue;
+		}
+
+		if (++*num >= max_num)
+			break;
+	}
+
+	closedir(dir);
+	if (!*num) {
+		infof("No suitable input events devices found.");
+		return -1;
+	}
+
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+sysfs_write_int(const char *fname, int val)
+{
+	FILE *f;
+
+	if (!(f = fopen(fname, "w"))) {
+		errorf("Can't open \"%s\" for writing", fname);
+		return -1;
+	}
+
+	fprintf(f, "%d\n", val);
+	fclose(f);
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int is_display_on = 1;
+
+static int
+turn_display_on(void)
+{
+	if (is_display_on)
+		return 0;
+
+	debugf("Turning display on.");
+	is_display_on = 1;
+	return sysfs_write_int(DISPLAY_CONTROL, 0);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+turn_display_off(void)
+{
+	if (!is_display_on)
+		return 0;
+
+	debugf("Turning display off.");
+	is_display_on = 0;
+	return sysfs_write_int(DISPLAY_CONTROL, 1);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+handle_events(int fd)
+{
+	int rv;
+	char buf[EVENTS_BUF_SIZE * sizeof(struct input_event)];
```

Comment:
Why not just: struct input_event buf[EVENTS_BUF_SIZE]?

Does not really matter now as the events are not actually handled, but if that changes later on using char buffer allows compiler to use alignment that might not be usable as input_event array.